sbt_common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/ElasticsearchIndexCreator.scala

```diff
@@ -0,0 +1,74 @@
+package uk.ac.wellcome.elasticsearch
+
+import com.sksamuel.elastic4s.http.{ElasticClient, Response}
+import com.sksamuel.elastic4s.http.ElasticDsl.{createIndex, _}
+import com.sksamuel.elastic4s.http.index.CreateIndexResponse
+import com.sksamuel.elastic4s.http.index.mappings.PutMappingResponse
+import com.sksamuel.elastic4s.mappings.MappingDefinition
+import com.sksamuel.elastic4s.mappings.dynamictemplate.DynamicMapping
+import grizzled.slf4j.Logging
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class ElasticsearchIndexCreator(elasticClient: ElasticClient)(
+  implicit ec: ExecutionContext)
+    extends Logging {
+  def create(indexName: String,
+             mappingDefinitionBuilder: MappingDefinitionBuilder): Future[Unit] =
+    create(
+      indexName = indexName,
+      mappingDefinition =
+        mappingDefinitionBuilder.buildMappingDefinition(indexName)
+    )
+
+  private def create(indexName: String,
+                     mappingDefinition: MappingDefinition): Future[Unit] =
+    elasticClient
+      .execute {
+        createIndex(indexName).mappings {
+          mappingDefinition
+        }
+      }
+      .map { response: Response[CreateIndexResponse] =>
+        if (response.isError) {
+          if (response.error.`type` == "resource_already_exists_exception" || response.error.`type` == "index_already_exists_exception") {
+            info(s"Index $indexName already exists")
+            update(indexName, mappingDefinition = mappingDefinition)
```

Comment:
Doesn't this return a Future[Future[_]]? are we sure we wait on the result of the uopdate? Is there a test for it?

Reply:
I don't think there's a test that we wait for it, and I'm not sure we can write one – the nature of Elasticsearch indexes is that sometimes they take a while even when the API completes. We do have a test that it successfully updates a mapping, but there's a 500ms sleep first.

But it is returning `Future[Future[Unit]]`; good catch. Will fix.